core/src/main/java/org/apache/iceberg/avro/ValueWriters.java

```diff
@@ -492,4 +576,215 @@ protected Object get(IndexedRecord struct, int pos) {
+    }
+
+    @Override
+    public Stream<FieldMetrics> metrics() {
+      // convert min/max to byte buffer to allow upper/lower bound truncation when gathering metrics.
+      return metrics(ByteBuffer::wrap);
+    }
+  }
+
+  public abstract static class ComparableWriter<T extends Comparable<T>> extends MetricsAwareWriter<T> {
+    public ComparableWriter(int id) {
+      super(id, Comparable::compareTo);
+    }
+  }
+
+  /**
+   * A value writer wrapper that keeps track of column statistics (metrics) during writing.
+   *
+   * @param <T> Input type
+   */
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  public abstract static class MetricsAwareWriter<T> implements ValueWriter<T> {
+    protected final int id;
+    protected long valueCount;
+    protected T max;
+    protected T min;
+
+    private final Comparator<T> comparator;
+
+    public MetricsAwareWriter(int id, Comparator<T> comparator) {
+      this.id = id;
+      this.comparator = comparator;
+    }
+
+    @Override
+    public void write(T datum, Encoder encoder) throws IOException {
+      valueCount++;
+
+      // null value should be handled by option writer, thus assume datum will not be null here.
+      if (max == null || comparator.compare(datum, max) > 0) {
+        max = datum;
+      }
+
+      if (min == null || comparator.compare(datum, min) < 0) {
+        min = datum;
+      }
+
+      writeVal(datum, encoder);
+    }
+
+    protected abstract void writeVal(T datum, Encoder encoder) throws IOException;
+
+    @Override
+    public Stream<FieldMetrics> metrics() {
+      return Stream.of(new FieldMetrics<>(id, valueCount, 0, 0, min, max));
+    }
+
+    /**
+     * Helper class to transform the input type when collecting metrics.
+     * The transform function converts the stats information from the specific type that the underlying writer
+     * understands to a more general type that could be transformed to binary following iceberg single-value
+     * serialization spec.
+     *
+     * @param func transformation function
+     * @return a stream of field metrics with bounds converted by the given transformation
+     */
+    protected Stream<FieldMetrics> metrics(Function<T, ?> func) {
+      return Stream.of(new FieldMetrics<>(id, valueCount, 0, 0,
+          min == null ? null : func.apply(min), max == null ? null : func.apply(max)));
+    }
+  }
+
+  /**
+   * A value writer wrapper that keeps track of column statistics (metrics) during writing.
+   * Implementations have to supply a convert() function, which will be applied to the input
+   * data to produce the integer type that the underlying writer accepts. Stats will also be
+   * tracked with int type.
+   *
+   * @param <T> Input type before conversion
+   */
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  public abstract static class StoredAsIntWriter<T> implements ValueWriter<T> {
+    protected final int id;
+    protected long valueCount;
+    protected Integer max;
```

Comment:
Because this is only used for non-null values, I think you can simplify the min and max updates by removing the null check. Just use `protected int min = Integer.MAX_VALUE` here and remove the null check. Then when creating `FieldMetrics`, you can check whether `valueCount` is non-zero to know whether `min` and `max` are valid.